### ISSUE TITLE
<variable> used only once: possible typo

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9402]
Migrated from [rt.perl.org#56530](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=56530) (status was 'resolved')

Searchable as RT56530$

**Comment by p5pRT** [comment]
### From toralf.foerster@gmx.de

The following one\-liner shouldn't produce that warning\, isn't it ? :

tfoerste@&#8203;n22 ~ $ seq 1 7 | perl \-we 'while \(\<STDIN>\) \{ print "line break\\n" if \($counter\+\+ % 3 == 0\); print \}'
Name "main&#8203;::counter" used only once&#8203;: possible typo at \-e line 1\.
line break
1
2
3
line break
4
5
6
line break
7
tfoerste@&#8203;n22 ~ $ perl \-v

This is perl\, v5\.8\.8 built for i686\-linux

Copyright 1987\-2006\, Larry Wall

\-\- 
MfG/Sincerely

Toralf Förster
pgp finger print&#8203;: 7B1A 07F4 EC82 0F90 D4C2 8936 872A E508 7DB6 9DA3

**Comment by p5pRT** [comment]
### From @smpeters

On Wed\, Jul 2\, 2008 at 8&#8203;:23 AM\, via RT Toralf Förster
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  Toralf Förster
> \# Please include the string&#8203;:  \[perl \#56530\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=56530 >
>
>
> The following one\-liner shouldn't produce that warning\, isn't it ? :
>
> tfoerste@&#8203;n22 ~ $ seq 1 7 | perl \-we 'while \(\<STDIN>\) \{ print "line break\\n" if \($counter\+\+ % 3 == 0\); print \}'
> Name "main&#8203;::counter" used only once&#8203;: possible typo at \-e line 1\.
> line break
> 1
> 2
> 3
> line break
> 4
> 5
> 6
> line break
> 7

If you do not "use strict"\, then yes\, that error looks weird\, but it
is true\.  $counter is used in one spot in the program\.  use'ing strict
requires some changes that remove the warning\.

steve@&#8203;picard&#8203;:\~/$ seq 1 7 | perl \-Mstrict \-we 'my $counter = 0; while
\(\<STDIN>\) \{ print "line break\\n" if \($counter\+\+ % 3 == 0\); print \}'
line break
1
2
3
line break
4
5
6
line break
7

Steve Peters
steve@&#8203;fisharerojo\.org

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
p5p@spam.wizbit.be - Status changed from 'open' to 'resolved'

**Comment by p5pRT** [comment]
### From toralf.foerster@gmx.de

At Friday 04 July 2008 23&#8203;:11&#8203;:53 Bram via RT wrote :
> According to our records\, your request regarding 
>   "\<variable> used only once&#8203;: possible typo" 
> has been resolved\. 
> 
> If you have any further questions or concerns\, please respond to this message\.
> 
> For other topics\, please create a new ticket\.
> 
> Please don't feel obligated to say "Thanks" or "Kudos" or "I owe you a beer" \-\- if you respond to this message it will reopen the ticket\. If you must\, please send email directly to the person who handled your ticket\, and not to the tracking system\.
> 
> \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=56530 >
> 

Yes\, "perl works as designed" might be an answer\.
OTOH I'm wondering whether perl could be improved to not give a warning in such a case\.
The term "used only once" is true for a line like "$foo = 1; " if ther's no further access to
the variable foo\, but the given example shows that the variable is used after it is incremented\.\.\.


\-\- 
MfG/Sincerely

Toralf Förster
pgp finger print&#8203;: 7B1A 07F4 EC82 0F90 D4C2 8936 872A E508 7DB6 9DA3

**Comment by p5pRT** [comment]
### From p5p@perl.wizbit.be

Citeren Toralf Förster \<toralf\.foerster@&#8203;gmx\.de>&#8203;:

> At Friday 04 July 2008 23&#8203;:11&#8203;:53 Bram via RT wrote :
>> According to our records\, your request regarding
>>   "\<variable> used only once&#8203;: possible typo"
>> has been resolved\.
>>
>> If you have any further questions or concerns\, please respond to   
>> this message\.
>>
>> For other topics\, please create a new ticket\.
>>
>> Please don't feel obligated to say "Thanks" or "Kudos" or "I owe   
>> you a beer" \-\- if you respond to this message it will reopen the   
>> ticket\. If you must\, please send email directly to the person who   
>> handled your ticket\, and not to the tracking system\.
>>
>> \<URL&#8203;: http&#8203;://rt\.perl\.org/rt3/Ticket/Display\.html?id=56530 >
>>
>
> Yes\, "perl works as designed" might be an answer\.
> OTOH I'm wondering whether perl could be improved to not give a   
> warning in such a case\.
> The term "used only once" is true for a line like "$foo = 1; " if   
> ther's no further access to
> the variable foo\, but the given example shows that the variable is   
> used after it is incremented\.\.\.
>

What really counts is how many times the variable is written\, which is  
indeed not always the same as used\.

In the case of your example&#8203;: $counter is written only once in the  
script\. The value is accessed and then incremented in a single  
operation\.
This might be an error\, this might not be an error\.

What if your example was for example&#8203;:

seq 1 7 | perl \-we '$counter = 1; while \(\<STDIN>\) \{ print "line  
break\\n" if \($couter\+\+ % 3 == 0\); print "line $counter = $\_" \}'


Personally\, I don't really see a value \(or a way\) in imporoving it\.\.\.


You don't want to see the warning?
Use a lexical variable\.

You don't want to bother use a lexical variable?
Then disable warnings\.


Warnings are about things that could be wrong\.
If you write the name of a global variable only once in your script  
then that certainly could be wrong\.


Kind regards\,

Bram
